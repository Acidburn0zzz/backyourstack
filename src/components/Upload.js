import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Dropzone from 'react-dropzone';
import fetch from 'cross-fetch';

import { dependenciesStats } from '../lib/utils';

export default class Upload extends React.Component {

  static propTypes = {
    files: PropTypes.object,
    onUpload: PropTypes.func,
    onUpdate: PropTypes.func,
  };

  static defaultProps = {
    files: {},
  };

  onDrop = (acceptedFiles) => {
    if (acceptedFiles.length > 0) {
      const formData = new FormData();
      acceptedFiles.forEach(file => {
        formData.append('files', file);
      });
      fetch('/files/upload', { method: 'POST', body: formData })
        .then(() => {
          if (this.props.onUpload) {
            this.props.onUpload();
          }
        });
    }
  };

  handleRemoveFile = (id, event) => {
    event.stopPropagation();
    const params = {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ id }),
    };
    fetch('/files/delete', params)
      .then(() => {
        if (this.props.onUpdate) {
          this.props.onUpdate();
        }
      });
  };

  render () {
    const fileEntries = Object.entries(this.props.files);

    return (
      <Fragment>
        <style jsx global>{`
        .dropZoneArea {
          border-width: 1px;
          border-color: #9399A3;
          border-style: dashed;
          border-radius: 4px;
          min-height: 125px;
          position: relative;
          color: #9399A3;
          font-size: 12px;
        }
        .dropZoneArea .empty {
          position: absolute;
          width: 100%;
          top: 40%;
          text-align: center;
        }
        .Files {
          display: flex;
          flex-wrap: wrap;
          justify-content: center;
        }
        .File {
          width: 180px;
          border-radius: 15px;
          background-color: #ffffff;
          box-shadow: 0 1px 3px 0 rgba(45, 77, 97, 0.2);
          margin: 10px;
          padding: 15px;
          text-align: center;
        }
        .File .name {
          height: 50px;
        }
        `}
        </style>
        <Dropzone onDrop={this.onDrop} className="dropZoneComponent">
          <div className="dropZoneArea">
            {fileEntries.length === 0 &&
              <p className="empty">
                Simply drag&#39;n&#39;drop files or click to select files to upload.
              </p>
            }
            {fileEntries.length > 0 &&
              <div className="Files">
                {fileEntries.map(([ id, file ]) => (
                  <div key={id} className="File">
                    <div className="name">
                      <b>{file.parsed.name || 'Unnamed project'}</b>
                    </div>
                    <div className="dependencies">
                      <b>{dependenciesStats(file.parsed).length}</b> dependencies
                    </div>
                    <div className="removefile">
                      <button onClick={(event) => this.handleRemoveFile(id, event)}>Remove</button>
                    </div>
                  </div>
                ))}
              </div>
            }
          </div>
        </Dropzone>
      </Fragment>
    );
  }

}
