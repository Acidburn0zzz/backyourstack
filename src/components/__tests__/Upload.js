import React from 'react';
import Upload from '../Upload';
import renderer from 'react-test-renderer';

test('Upload should render', () => {
  const component = renderer.create(<Upload />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Upload should render with feedbackPosition', () => {
  const component = renderer.create(<Upload feedbackPosition="inside" />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Upload should render with style', () => {
  const component = renderer.create(
    <Upload style={{ border: '1px solid red' }} />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
