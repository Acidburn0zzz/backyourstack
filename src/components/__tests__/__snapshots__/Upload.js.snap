// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload should render 1`] = `
<div
  aria-disabled={false}
  className="dropZoneComponent"
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDragStart={[Function]}
  onDrop={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="jsx-47610240 text"
  >
    <p
      className="jsx-47610240"
    >
      Simply drag'n'drop files
      <br
        className="jsx-47610240"
      />
      or click to select files to upload.
    </p>
  </div>
  <div
    className="jsx-47610240 uploadFeedback float"
  >
    <p
      className="jsx-47610240"
    >
      There was an error while uploading your files. At the moment, we do support 
      <em>
        composer.json
      </em>
      , 
      <em>
        *.csproj
      </em>
      , 
      <em>
        packages.config
      </em>
      , 
      <em>
        Gopkg.lock
      </em>
       and 
      <em>
        Gemfile.lock
      </em>
      . Please try again. If the problem persists, please contact us.
    </p>
  </div>
  <input
    autoComplete="off"
    disabled={false}
    multiple={true}
    onChange={[Function]}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "opacity": 0.00001,
        "pointerEvents": "none",
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    type="file"
  />
</div>
`;

exports[`Upload should render with feedbackPosition 1`] = `
<div
  aria-disabled={false}
  className="dropZoneComponent"
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDragStart={[Function]}
  onDrop={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="jsx-47610240 text"
  >
    <p
      className="jsx-47610240"
    >
      Simply drag'n'drop files
      <br
        className="jsx-47610240"
      />
      or click to select files to upload.
    </p>
  </div>
  <div
    className="jsx-47610240 uploadFeedback inside"
  >
    <p
      className="jsx-47610240"
    >
      There was an error while uploading your files. At the moment, we do support 
      <em>
        composer.json
      </em>
      , 
      <em>
        *.csproj
      </em>
      , 
      <em>
        packages.config
      </em>
      , 
      <em>
        Gopkg.lock
      </em>
       and 
      <em>
        Gemfile.lock
      </em>
      . Please try again. If the problem persists, please contact us.
    </p>
  </div>
  <input
    autoComplete="off"
    disabled={false}
    multiple={true}
    onChange={[Function]}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "opacity": 0.00001,
        "pointerEvents": "none",
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    type="file"
  />
</div>
`;

exports[`Upload should render with style 1`] = `
<div
  aria-disabled={false}
  className="dropZoneComponent"
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDragStart={[Function]}
  onDrop={[Function]}
  style={
    Object {
      "border": "1px solid red",
      "position": "relative",
    }
  }
>
  <div
    className="jsx-47610240 text"
  >
    <p
      className="jsx-47610240"
    >
      Simply drag'n'drop files
      <br
        className="jsx-47610240"
      />
      or click to select files to upload.
    </p>
  </div>
  <div
    className="jsx-47610240 uploadFeedback float"
  >
    <p
      className="jsx-47610240"
    >
      There was an error while uploading your files. At the moment, we do support 
      <em>
        composer.json
      </em>
      , 
      <em>
        *.csproj
      </em>
      , 
      <em>
        packages.config
      </em>
      , 
      <em>
        Gopkg.lock
      </em>
       and 
      <em>
        Gemfile.lock
      </em>
      . Please try again. If the problem persists, please contact us.
    </p>
  </div>
  <input
    autoComplete="off"
    disabled={false}
    multiple={true}
    onChange={[Function]}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "opacity": 0.00001,
        "pointerEvents": "none",
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    type="file"
  />
</div>
`;
